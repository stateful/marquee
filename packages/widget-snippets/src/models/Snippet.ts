import path from 'path'
import { v4 as uuidv4 } from 'uuid'
import vscode from 'vscode'

export default class Snippet implements vscode.FileStat {
  type: vscode.FileType = vscode.FileType.File
  size: number
  path: string
  origin?: string
  ctime: number
  mtime: number

  constructor (
    public workspaceId: string | null = null,
    public title = 'New Snippet',
    public body = '',
    public archived = false,
    public createdAt = Date.now(),
    public id = uuidv4(),
    snippetPath?: string
  ) {
    this.type = vscode.FileType.File
    this.size = body.length
    this.path = path.join(`/${id}`, snippetPath ? path.basename(snippetPath) : title)
    this.origin = snippetPath
    this.ctime = this.createdAt
    this.mtime = this.createdAt
  }

  static fromObject (s: Partial<Snippet>) {
    const snippet = new Snippet(s.workspaceId, s.title, s.body, s.archived, s.createdAt, s.id, s.origin)
    return snippet
  }

  get data () {
    const enc = new TextEncoder()
    return enc.encode(this.body)
  }
}
