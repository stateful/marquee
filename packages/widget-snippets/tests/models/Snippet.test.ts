import path from 'path'
import Snippet from '../../src/models/Snippet'

describe('Snippet', () => {
  it('can be created with defaults', () => {
    const s = new Snippet()
    expect(s.workspaceId).toBe(null)
    expect(s.title).toBe('New Snippet')
    expect(s.path).toBe(path.sep + path.join('39d2f858-3522-4053-beb8-b75a1defd1d2', 'New Snippet'))
  })

  it('can be created from snippet object', () => {
    const s = Snippet.fromObject({
      id: 'foobar',
      title: 'title',
      archived: false,
      body: 'some body',
      workspaceId: null,
      createdAt: 123
    })
    // eslint-disable-next-line
    const { path, ...snippet } = s
    expect(snippet).toMatchSnapshot()
    expect(s.data).toMatchSnapshot()
  })
})
