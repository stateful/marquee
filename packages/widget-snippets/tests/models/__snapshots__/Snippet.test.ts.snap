// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snippet can be created from snippet object 1`] = `
Snippet {
  "archived": false,
  "body": "some body",
  "createdAt": 123,
  "ctime": 123,
  "id": "foobar",
  "mtime": 123,
  "origin": undefined,
  "path": "/foobar/title",
  "size": 9,
  "title": "title",
  "type": "1",
  "workspaceId": null,
}
`;

exports[`Snippet can be created from snippet object 2`] = `
Uint8Array [
  115,
  111,
  109,
  101,
  32,
  98,
  111,
  100,
  121,
]
`;
